### DISCUSSION TITLE
Caprover / Pocketbase Persistent Data

### DISCUSSION THREAD

**Discussion by onlyspike** [pocketbase/pocketbase#3080] [Q&A] [answered]
Hi,
I have recently discovered Caprover and have set up a little server over the weekend.
I had some issues with the One-Click App process for Pocketbase, but I was able to finally get it up and running with a Dockerfile.
I currently use Github Actions to deploy the changes to Caprover.
The issue I am having is each time I make a change in my "pb_hooks" and do a commit, the "pb_data" is cleared.
I hope someone would be able to assist on a process on how I can update my "pb_hooks" changes without having the "pb_data" being cleared.
I have set up the app in Caprover to use Persistent Data, but I am not able to get it to work.
This is might be a question that should be posted on Caprover's discussion, but I am hoping this someone here have been able to get this working.

Below is my Dockerfile.

```
FROM alpine:latest

RUN apk add -v build-base
RUN apk add -v go 
RUN apk add -v ca-certificates
RUN apk add --no-cache \
    unzip \
    # this is needed only if you want to use scp to copy later your pb_data locally
    openssh

COPY ./gocode /pb
COPY ./pb_hooks /pb/pb_hooks
WORKDIR /pb
RUN go mod tidy -e

RUN go get github.com/pocketbase/pocketbase

RUN go build
WORKDIR /

EXPOSE 5015

CMD ["/pb/pocketbase", "serve", "--http=0.0.0.0:5015"]
```

Much appreciated.

**Comment by ganigeorgiev** [comment]
I'm not familiar with CapRover and I'm not sure that I'll be able to help you but double check that `/pb/pb_data` is actually mounted as persistent volume in your container.

I'm also a little confused with the above Dockerfile. Why do you need to manually build PocketBase with Go if you are using the JS hooks? Or do you use both JS and Go hooks in your app?

**Reply by ganigeorgiev** [reply]
_From the [CapRover docs](https://caprover.com/docs/persistent-apps.html#persistent-apps):_

> Note that even for Persistent Apps, NOT ALL DIRECTORIES will be treated as persistent directories. After you created the app as an app with persistent data, **you'll have to define directories that you want to be persistent in the app details page on web dashboard**. You can let CapRover manage the stored directories for you (use labels), or use a specific path on the host (server).

**Reply by onlyspike** [reply]
Hi Gani,

Thank you for your response.

The reason why I am using the Go build is thats the only way I got it working.
When I tried the suggested Docker process from Pocketbase or when using the One-Click Apps process on Caprover, I keep getting an error : "exec /pb/pocketbase: exec format error"

Yes, I have set the Persistent Directories and they do exist on the server.

**Reply by ganigeorgiev** [reply]
It shouldn't be necessary to manually build PocketBase if you are not going to extend it with Go (or want a CGO build) and it would just increases the image size and slow the build process.

The reason for the "exec format error" is that the Caprover One-Click app is probably using an executable for different platform (see also a recent PR in CapRover that allow specifying a different architecture - https://github.com/caprover/one-click-apps/pull/940).

Unfortunately I have no idea why the persistent directories don't work for you. You could try to ask the question in the CapRover related forum to see if someone has better idea what could be the cause of the issue.

**Reply by onlyspike** [reply]
When using the One-Click progress it was defaulted to 'linux-amd64'. I tried 'linux-arm64' and that worked.
I just now need to still figure out how to push my pb_hooks files and have the data persist.
I'll post a discussion on CapRover as well.

**Reply by xaqbr** [reply, marked as answer]
@onlyspike My data seems to be persisting OK with the custom Dockerfile option (i only changed the ports):

```dockerfile
FROM alpine:latest

ARG PB_VERSION=0.16.10

RUN apk add --no-cache \
    unzip \
    ca-certificates

# download and unzip PocketBase
ADD https://github.com/pocketbase/pocketbase/releases/download/v${PB_VERSION}/pocketbase_${PB_VERSION}_linux_amd64.zip /tmp/pb.zip
RUN unzip /tmp/pb.zip -d /pb/

# uncomment to copy the local pb_migrations dir into the image
COPY ./pb_migrations /pb/pb_migrations

# uncomment to copy the local pb_hooks dir into the image
COPY ./pb_hooks /pb/pb_hooks

EXPOSE 80

# start PocketBase
CMD ["/pb/pocketbase", "serve", "--http=0.0.0.0:80"]
```

Here's a pic of my persistent directory mappings
<img width="797" alt="Screenshot 2023-08-07 at 2 50 55 PM" src="https://github.com/pocketbase/pocketbase/assets/3822492/e30465a4-4a04-47b1-abfe-5e2efb19c7ed">

**Reply by onlyspike** [reply]
Hi Jaywack,

Thank you for your assistance. This was exactly what I was looking for.

A couple of things I had to do on my side.
On Dockerfile. I changed "linux_amd64" to "linux_arm64"

I also removed the persistent directory for "pb_hooks". I found that when I pushed my changes it would not update, but by removing the folder as a persistent directory it would use the latest files I commit.

Thank you for everyone's assistance.
